### PR TITLE
#[166240827] Users should be able to give a rating to an article

### PR DESCRIPTION
**Pull request by MCFrank16**
#### Usecase
[usecase link](https://docs.google.com/document/d/1KONxvg4D5BQ7BkSJo56y7YrgDQcAeXJe-yf77VYNoko/edit)
#### What does this PR do?

this PR shows how the authors will be able to update his/her article.

#### Description of Task to be completed?

Have the following endpoint working: 
POST ``/api/articles/:slug/rating``: this endpoint will allow a user/reader to rate a particular slug based on its slug.

#### How should this be manually tested?
* Clone the repo to  local machine.
* Run ``npm install``   to install the dependencies.
* Set up the   ``.env`` as instructed by the   ``.env-example``
* Run  ``npm  run dev``.
* POST the rating data using the following endpoint  ``/api/articles/:slug/rating``.

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[166240827](https://www.pivotaltracker.com/story/show/166240827)

### CODE REVIEW

File: src/api/controllers/rateArticleController.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    const { id } = user;
+
+    try {
+      await Ratings.create({
+        userId: id,
+        articleSlug: slug,
+        rating: parseInt(Rating, 10)
+      });
+
+      return res.status(statuses.OK).json({
```
Comment (alainburindi) [inline src/api/controllers/rateArticleController.js]: what do you think about 200 as status on create?

File: src/middlewares/getOneArticle.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+   * @memberof getOneArticle
+   * @returns {Object} res
+   */
+  static async getArticle(req, res, next) {
+    const { slug } = req.params;
+
+    const result = await Article.findOne({
+      where: {
+        slug,
+      }
```
Comment (alainburindi) [inline src/middlewares/getOneArticle.js]: where do you check if the user is the owner of the article?

File: tests/rateArticle.test.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      })
+      .end((err, res) => {
+        userToken = res.body.user.token;
+        done();
+      });
+  });
+  it('should login a user and get a token to use against on rating article', (done) => {
+    chai
+      .request(app)
+      .post('/api/users/login')
```
Comment (alainburindi) [inline tests/rateArticle.test.js:14]: why login and yet you have the token on register??

File: src/api/controllers/rateArticleController.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        status: statuses.OK,
+        message: `you rated ${slug} with ${Rating} rates, Thanks`
+      });
+    } catch (err) {
+      sendError(statuses.BAD_REQUEST,
+        res,
+        'Rating',
+        'Sorry, you can not update this article twice with the same rating number, please update it');
+    }
```
Comment (alainburindi) [inline src/api/controllers/rateArticleController.js]: are you sure this is the correct message?? you used update

File: src/api/controllers/rateArticleController.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        userId: id,
+        articleSlug: slug,
+        rating: parseInt(Rating, 10)
+      });
+
+      return res.status(statuses.OK).json({
+        status: statuses.OK,
+        message: `you rated ${slug} with ${Rating} rates, Thanks`
+      });
+    } catch (err) {
```
Comment (alainburindi) [inline src/api/controllers/rateArticleController.js]: what is the limit or max value of rating?

**Review by alainburindi — CHANGES_REQUESTED** [review]

File: src/api/controllers/rateArticleController.js (hunk 1 above)
Comment (MCFrank16) [reply src/api/controllers/rateArticleController.js]: it's a 201 my bad

File: src/middlewares/getOneArticle.js (hunk 2 above)
Comment (MCFrank16) [reply src/middlewares/getOneArticle.js]: there is another middleware for that #checkarticleOwnership

File: src/api/controllers/rateArticleController.js (hunk 4 above)
Comment (MCFrank16) [reply src/api/controllers/rateArticleController.js]: yeah I think the user needs to update the rating number and remove the old one

File: src/api/controllers/rateArticleController.js (hunk 5 above)
Comment (MCFrank16) [reply src/api/controllers/rateArticleController.js]: so far I didn't think about it? what's your suggestion @alainburindi

File: tests/rateArticle.test.js (hunk 3 above)
Comment (MCFrank16) [reply tests/rateArticle.test.js:14]: Yeah you are right there is no reason to have both, I am on it now

File: src/api/controllers/rateArticleController.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
+import model from '../models';
```
Comment (premlogic) [inline src/api/controllers/rateArticleController.js]: We agreed on the name of this file together @MCFrank16. it should be ```articleRatingController```

File: src/api/controllers/rateArticleController.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+   * @static
+   * @param {object} req - this is the request object
+   * @param {object} res - this is the response object
+   * @returns {object} - this function returns an object
+   * @memberof RatingsController
+   */
+  static async rateArticle(req, res) {
+    const { slug } = req.params;
+    const { rating } = req.body;
+    const { user } = req.user;
```
Comment (premlogic) [inline src/api/controllers/rateArticleController.js]: you can improve this line like this ```{ user: { id } } = user```

File: src/api/routes/articleRouter.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import { Router } from 'express';
+import checkArticle from '../../middlewares/getOneArticle';
+import checkValidToken from '../../middlewares/checkValidToken';
+import bodyVerifier from '../../middlewares/validations/body.verifier';
+import ratingsController from '../controllers/rateArticleController';
+import validateInputs from '../../middlewares/validations/body.inputs';
+
+const articleRouter = new Router();
+
+articleRouter.post('/:slug/rating',
```
Comment (premlogic) [inline src/api/routes/articleRouter.js:15]: check also if the body is empty. I created a middleware for that. @MCFrank16

File: src/helpers/constants/numbers.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+export default {
```
Comment (premlogic) [inline src/helpers/constants/numbers.js:1]: Wow.. Finally. We're gonna use this one @MCFrank16

**Review by nkpremices — CHANGES_REQUESTED** [review]

File: src/api/routes/articleRouter.js (hunk 8 above)
Comment (MCFrank16) [reply src/api/routes/articleRouter.js:15]: I did it man

File: src/api/controllers/rateArticleController.js (hunk 7 above)
Comment (MCFrank16) [reply src/api/controllers/rateArticleController.js]: yeah this is the best way

File: src/api/controllers/rateArticleController.js (hunk 6 above)
Comment (MCFrank16) [reply src/api/controllers/rateArticleController.js]: yeah lemme change it

**Comment by MemunaHaruna** [comment]
@MCFrank16 please, refer to the engineering playbook on how to properly name branches. No capital letters and no #. Also, squash commits on your PR.

File: src/api/controllers/articleRatingController.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
+import model from '../models';
+import statuses from '../../helpers/constants/status.codes';
+import sendError from '../../helpers/error.sender';
+
+const { Ratings } = model;
```
Comment (MemunaHaruna) [inline src/api/controllers/articleRatingController.js]: @MCFrank16 Nitpick: let's use singular format for model names

File: src/api/controllers/articleRatingController.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    try {
+      await Ratings.create({
+        userId: id,
+        articleSlug: slug,
+        rating: parseInt(rating, 10)
+      });
+
+      return res.status(statuses.CREATED).json({
+        status: statuses.CREATED,
+        message: `you rated ${slug} with ${rating} rates, Thanks`
```
Comment (MemunaHaruna) [inline src/api/controllers/articleRatingController.js]: @MCFrank16 Change this text to `Rating for ${slug} submitted successfully`

File: src/api/migrations/20190621163915-create-article.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        model: 'Users',
+        key: 'id'
+      },
+      onDelete: 'CASCADE',
+      onUpdate: 'CASCADE',
+      allowNull: false
+    }
+  }),
+  down: (queryInterface, Sequelize) => queryInterface.dropTable('Articles')
+};
```
Comment (MemunaHaruna) [inline src/api/migrations/20190621163915-create-article.js]: @MCFrank16 add a newline here

File: src/api/migrations/20190621210101-create-sequelize-meta.js [hunk 13]
```diff
@@ -0,0 +1,8 @@
+export default {
```
Comment (MemunaHaruna) [inline src/api/migrations/20190621210101-create-sequelize-meta.js]: @MCFrank16 we no longer need this.

File: src/api/models/article.js [hunk 14]
```diff
@@ -0,0 +1,57 @@
+/* eslint-disable new-cap */
```
Comment (MemunaHaruna) [inline src/api/models/article.js]: @MCFrank16 no eslint disable

File: src/api/models/sequelizemeta.js [hunk 15]
```diff
@@ -0,0 +1,8 @@
+export default (sequelize, DataTypes) => {
```
Comment (MemunaHaruna) [inline src/api/models/sequelizemeta.js]: @MCFrank16 this isn't necessary

File: src/middlewares/getOneArticle.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
+import models from '../api/models';
```
Comment (MemunaHaruna) [inline src/middlewares/getOneArticle.js]: @MCFrank16 you don't need this in your PR

**Review by MemunaHaruna — APPROVED** [review]
@MCFrank16 we need to get create articles merged so that we know your PR only contains changes related to rating articles.

**Comment by MCFrank16** [comment]
> @MCFrank16 please, refer to the engineering playbook on how to properly name branches. No capital letters and no #. Also, squash commits on your PR.

yes I will from now on

File: src/api/controllers/articleRatingController.js (hunk 11 above)
Comment (MCFrank16) [reply src/api/controllers/articleRatingController.js]: yeah this is well structured

File: src/middlewares/getOneArticle.js (hunk 16 above)
Comment (MCFrank16) [reply src/middlewares/getOneArticle.js]: I just use it to check if the article slug exists or not

File: src/api/controllers/articleRatingController.js [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+    const { rating } = req.body;
+    const { user: { id } } = req.user;
+
+    Rating.create({
+      userId: id,
+      articleSlug: slug,
+      rating: parseInt(rating, 10)
+    });
+    res.status(statuses.CREATED).json({
+      status: statuses.CREATED,
```
Comment (alainburindi) [inline src/api/controllers/articleRatingController.js]: I think you talked with @truestbyheart about statuses vs status vs statusCodes

File: src/api/migrations/20190622195903-add-unique-constraint-article.js [hunk 18]
```diff
@@ -0,0 +1,7 @@
+export default {
+  up: (queryInterface, Sequelize) => queryInterface.addConstraint('Articles', ['slug'], {
```
Comment (MemunaHaruna) [inline src/api/migrations/20190622195903-add-unique-constraint-article.js:2]: cool

**Review by alainburindi — CHANGES_REQUESTED** [review]

File: tests/updateArticle.test.js [hunk 19]
```diff
@@ -1,8 +1,8 @@
 import chai from 'chai';
 import chaiHttp from 'chai-http';
 import app from '../src/index';
-import slugMaker from '../src/helpers/slug.maker';
-import statuses from '../src/helpers/constants/status.codes';
+// import slugMaker from '../src/helpers/slug.maker';
```
Comment (MemunaHaruna) [inline tests/updateArticle.test.js]: @MCFrank16 remove unused commented code

Comment (MCFrank16) [reply tests/updateArticle.test.js]: Yeah, I have removed it

**Pull request merged**